### PR TITLE
Add cover property API that can be refined through Config PropertyLibrary

### PR DESCRIPTION
**Pull request by jchang0**
Also make variable to TLToAXI4 public

### CODE REVIEW

File: src/main/scala/util/Property.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  override def toString: String = name
+}
+
+object PropertyType {
+  object Assert extends PropertyType("Assert")
+  object Assume extends PropertyType("Assume")
+  object Cover extends PropertyType("Cover")
+}
+
+class BasePropertyParameters(val p_type: PropertyType, val cond: Bool, val label: String, val message: String) {}
```
Comment (hcook) [inline src/main/scala/util/Property.scala]: i would turn this into a trait (can be extended by case classes):
```
trait BasePropertyParameters {
  val pType: PropertyType
  val cond: Bool
  val label: String
  val message: String
}

case class CoverPropertyParameters(
    cond: Bool,
    label: String = "",
    message: String = "") extends BasePropertyParameters {
  val pType =  PropertyType.Cover
}
```

File: src/main/scala/util/Property.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+class BasePropertyLibrary[PropParam <: BasePropertyParameters] {
+  def generateCover(prop_param: PropParam)(implicit sourceInfo: SourceInfo): Unit = {
+    // Currently empty for user to extend their own cover methods
+    Unit
+  }
+}
```
Comment (hcook) [inline src/main/scala/util/Property.scala]: i would turn this into an abstract class:
```
abstract class BasePropertyLibrary[PP <: BasePropertyParameters] {
  def generateCover(pp: PP)(implicit sourceInfo: SourceInfo): Unit
}
```
and then maybe add a default concrete class that does something more specific:
```
class PrintCoverPropertyLibrary extends BasePropertyLibrary[CoverPropertyParameters] {
   def generateCover(pp: CoverPropertyParameters)(implicit sourceInfo: SourceInfo) { printf(p.message) }
}
```

File: src/main/scala/util/Property.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  }
+  def apply(cond: Bool, label: String)(implicit sourceInfo: SourceInfo, p: Parameters): Unit = {
+    val prop_param = new BasePropertyParameters(PropertyType.Cover, cond, label, "")
+    p(PropertyLibrary).generateCover(prop_param)
+  }
+  def apply(cond: Bool, label: String, message: String)(implicit sourceInfo: SourceInfo, p: Parameters): Unit = {
+    val prop_param = new BasePropertyParameters(PropertyType.Cover, cond, label, message)
+    p(PropertyLibrary).generateCover(prop_param)
+  }
+}
```
Comment (hcook) [inline src/main/scala/util/Property.scala]: If all `PropertyLibrary` child classes can agree that `generateCover` takes a `CoverPropertyParamers` then you can simplify this to the following
:
```
object cover {
  def apply(cond: Bool)(implicit sourceInfo: SourceInfo, p: Parameters) {
    p(PropertyLibrary).generateCover(CoverPropertyParamers(cond))
  }
  def apply(cond: Bool, label: String)(implicit sourceInfo: SourceInfo, p: Parameters) {
    p(PropertyLibrary).generateCover(CoverPropertyParamers(cond, label))
  }
  def apply(cond: Bool, label: String, message: String)(implicit sourceInfo: SourceInfo, p: Parameters) {
    p(PropertyLibrary).generateCover(CoverPropertyParamers(cond, label, message))
  }
}
```

File: src/main/scala/util/Property.scala (hunk 2 above)
Comment (hcook) [reply src/main/scala/util/Property.scala]: If you're returning Unit stylistically omit the `...): Unit = {...}` and just write `...) {...} `

Comment (hcook) [reply src/main/scala/util/Property.scala]: Should `generateCover` not admit instances of `BasePropertyParameters` where `p_type != Cover`?

File: src/main/scala/coreplex/Configs.scala [hunk 4]
```diff
@@ -27,6 +28,7 @@ class BaseCoreplexConfig extends Config ((site, here, up) => {
   case ErrorParams => ErrorParams(Seq(AddressSet(0x3000, 0xfff)))
   case BootROMParams => BootROMParams(contentFileName = "./bootrom/bootrom.img")
   case DebugModuleParams => DefaultDebugModuleParams(site(XLen))
+  case PropertyLibrary => (new BasePropertyLibrary[BasePropertyParameters])
```
Comment (hcook) [inline src/main/scala/coreplex/Configs.scala]: I wouldn't add a default here until we come up with a specific default behavior applied to instances of `cover(...)` in rocket-chip. See below...

**Comment by jchang0** [comment]
Made the requested changes.

**Comment by jackkoenig** [comment]
This looks good for federation and rocket-chip, but I'm wondering how we should approach it in a more general library that is not beholden to rocket-chips parameterization system. We don't have to solve that yet but just something to think about.

**Comment by jerryhethatday** [comment]
> This looks good for federation and rocket-chip, but I'm wondering how we should approach it in a more general library that is not beholden to rocket-chips parameterization system. We don't have to solve that yet but just something to think about.

Hey Jack, can you clarify in general what this `cover` object does? I saw many locs calling this `cover` functionality. But have no idea what this does. Thanks!

**Pull request merged**